libbeat/outputs/elasticsearch/client.go

```diff
@@ -48,7 +48,11 @@ var (
 
 // Client is an elasticsearch client.
 type Client struct {
-	conn eslegclient.Connection
+	// We use a pointer here because the connection holds a context
+	// that needs to be recreated in case of error and the connection
+	// that is passed to this client is also used in a closure, we need
+	// to ensure both hold a reference to the same instance of the connection.
```

Comment:
That you have to account for future usage at all here feels wrong too. We need to fix this in a way that does not require knowing future uses of the code. Make it impossible to have this bug again.

Why do we even need a closure, was it just there because it was convenient? Can we just not have a closure anymore? It looks like if we saved the captured `onConnect` it could be a method on the client to avoid also capturing the connection.

https://github.com/elastic/beats/blob/3c03c74b847bf35926a8b29aee6be76144982afe/libbeat/outputs/elasticsearch/client.go#L136-L160